presto-main/src/main/java/io/prestosql/execution/buffer/PagesSerde.java

```diff
@@ -56,86 +56,225 @@ public PagesSerde(BlockEncodingSerde blockEncodingSerde, Optional<Compressor> co
+
+                byte[] decrypted = context.acquireBuffer(spillCipher.get().decryptedMaxLength(slice.length()));
+                int decryptedSize = spillCipher.get().decrypt(
+                        slice.byteArray(),
+                        slice.byteArrayOffset(),
+                        slice.length(),
+                        decrypted,
+                        0);
+
+                slice = Slices.wrappedBuffer(decrypted, 0, decryptedSize);
+                inUseTempBuffer = decrypted;
+            }
+
+            if (serializedPage.isCompressed()) {
+                checkState(decompressor.isPresent(), "Page is compressed, but decompressor is missing");
+
+                int uncompressedSize = serializedPage.getUncompressedSizeInBytes();
+                byte[] decompressed = context.acquireBuffer(uncompressedSize);
+                checkState(decompressor.get().decompress(
+                        slice.byteArray(),
+                        slice.byteArrayOffset(),
+                        slice.length(),
+                        decompressed,
+                        0,
+                        uncompressedSize) == uncompressedSize);
 
-        if (serializedPage.isEncrypted()) {
-            checkState(spillCipher.isPresent(), "Page is encrypted, but spill cipher is missing");
+                slice = Slices.wrappedBuffer(decompressed, 0, uncompressedSize);
+                if (inUseTempBuffer != null) {
+                    //  Previous buffer is no longer in use
+                    context.releaseBuffer(inUseTempBuffer);
+                }
+                inUseTempBuffer = decompressed;
+            }
+
+            return readRawPage(serializedPage.getPositionCount(), slice.getInput(), blockEncodingSerde);
+        }
+        finally {
+            if (inUseTempBuffer != null) {
+                context.releaseBuffer(inUseTempBuffer);
+            }
+        }
+    }
 
-            byte[] decrypted = new byte[spillCipher.get().decryptedMaxLength(slice.length())];
-            int decryptedSize = spillCipher.get().decrypt(
-                    slice.byteArray(),
-                    slice.byteArrayOffset(),
-                    slice.length(),
-                    decrypted,
-                    0);
+    public static final class PagesSerdeContext
+            implements AutoCloseable
+    {
+        //  Limit retained buffers to 4x the default max page size
+        private static final int MAX_BUFFER_RETAINED_SIZE = DEFAULT_MAX_PAGE_SIZE_IN_BYTES * 4;
+
+        private DynamicSliceOutput sliceOutput;
+        //  Wraps two buffers since encryption + decryption will use at most 2 buffers at once. Buffers are kept in relative order
+        //  based on length so that they can be used for compression or encryption, maximizing reuse opportunities
+        private byte[] largerBuffer;
+        private byte[] smallerBuffer;
+        private boolean closed;
+
+        private void checkNotClosed()
```

Comment:
why do we need to close it? Java will automatically GC unused objects

Reply:
In some usages (eg: `PagesSerdeUtil#PageReader`) you would have to null out the `PagesSerdeContext` field instead of closing it to actually release the buffers, since you don't exactly know how long the iterator might still be reachable after reading the final element.

I think either approach could work, but implementing `AutoClosable` seemed like a  stronger signal about how the class should be used. If the context is kept for long periods of time then you have to start worrying about (and probably tracking) it's memory consumption which can be non-trivial.

Reply:
> since you don't exactly know how long the iterator might still be reachable after reading the final element.

The same applies to nullyfing buffers, they won't be released immediately.

> since you don't exactly know how long the iterator might still be reachable after reading the final element.

It's responsibility of iterator owner to release it after final element has been read.